Info Bar Directive

petzlux opened the pull request:
I have tried to copy the scaleline directive, and change it to an attribution control, to explore how the creation of custom controls works. Unfortunately, I cant seem to make it work, it doesn't give me any errors, but there is no attribution either ... 

Can someone point me in the right direction ?

elemoine commented:
Where are the attributions supposed to go? I do not see attributions on the [mockup](http://dev.camptocamp.com/files/lux/geoportail_lux_specs/cartes/). Will they be placed in the info bar together with the scaleline?

petzlux commented:
No its simply a test for me to see how adding a ol3 control through a directive works,  at the very least I would hope to see the console.log to execute,  but can't see that

elemoine commented on geoportailv3/static/js/infobardirective.js:
```diff
@@ -0,0 +1,58 @@
+};
+
+
+app.module.directive('appInfoBar', app.infobarDirective);
```
If you use `<app-infobar>` in the HTML then the name of the directive should be `appInfobar`. That's your error. Angular just ignores the DOM element, because there is no directive associated to it.

elemoine commented on geoportailv3/static/js/infobardirective.js:
```diff
@@ -0,0 +1,58 @@
+    restrict: 'E',
+    scope: {
+      'map': '=appInfoBarMap'
+    },
```
Use `appInfobarMap` here.

elemoine commented on geoportailv3/static/js/infobardirective.js:
```diff
@@ -0,0 +1,58 @@
+ * @ngInject
+ * @constructor
+ */
+app.InfoBarDirectiveController = function() {
```
`InfoBarDirectiveController` -> `InfobarDirectiveController` for consistency reasons here.

pgiraud commented on geoportailv3/static/js/infobardirective.js:
```diff
@@ -0,0 +1,107 @@
+    this['map'].addControl(this.mouseposition);
+
+    function _utmZoneCheck(coord){
+        var lonlat = ol.proj.transform(coord, $scope.projection.value, "EPSG:4326");
```
`$scope.projection.value` will equal "EPSG:3263*" which is not a valid projection.
You probably want to use "map.getView().getProjection()" instead.

petzlux replied:
see https://github.com/Geoportail-Luxembourg/geoportailv3/commit/218af1e20d85348eab20cc68bbf2d93de3122c03

its only a placeholder signifiying that we need to check which UTM band we should display  ...

petzlux commented:
@pgiraud & @elemoine : 

Yesterday we tried again to add proj4js as a dependency, to allow us to transform from EPSG:32362 / 32361 to EPSG:4326  (see 89c797b )

we added proj4 as a package in package.json , and added it as a src in build.json. We also added it in main.js and infobardirective.js as a goog.require('proj4') . NPM during build installs proj4, but the compiler then always complains that he cant find the package. 

This is now blocking finalisation of mouse coordinate display functionality. Can you please assist us in resolving this issue.

Thanks

pgiraud commented:
OK. I'm having a look.

pgiraud commented:
Is the code where you added proj4 to the dependencies pushed somewhere? I can't see it in this pull-request (ie. info_bar branch).

pgiraud commented:
So, proj4 need to be loaded in the page.
You'll need to add the following in `template.html`:
```html
<script src="${request.static_url('%s/proj4/dist/proj4-src.js' % node_modules_path)}"></script>
```

Also, you'll have to create the proj4 definition files.

```javascript
goog.provide('app.proj_2169');

goog.require('ol.Map');
goog.require('ol.proj');

proj4.defs('EPSG:2169',
    '+proj=tmerc +lat_0=49.83333333333334 +lon_0=6.166666666666667 ' +
    '+k=1 +x_0=80000 +y_0=100000 +ellps=intl ' +
    '+towgs84=-189.681,18.3463,-42.7695,-0.33746,-3.09264,2.53861,0.4598 ' +
    '+units=m +no_defs"');
ol.proj.get('EPSG:2169').setExtent([48225.17, 56225.60, 105842.04, 139616.40]);
```

Please pay attention to indentation and syntax errors in your code. It won't build in its current state.

elemoine commented:
What's important to note is that proj4 should be used as an external library. It should not be compiled with the application. ol3 already includes externs definitions for proj4.

rmichaelis commented:
Could someone explain me when we have to compile with the application and when we have to include as an external library ?

petzlux commented:
Thanks for these helpful comments, I am implementing right now and the problem is resolved !  Will push with changes to this pull request once finalised.

elemoine commented:
@rmichaelis, we compile code that can be compiled. As simple as that. In practice the app is compiled with OpenLayers 3, ngeo and Closure Library. For libraries that cannot be compiled (because their code is not compatible with Closure Compiler) we use externs definitions. The externs definitions provide types, and prevent renaming.

rmichaelis commented:
Okay, I understand now. Thanks for the information.

petzlux commented:
I have difficulties with the way closure compiler handles property values in https://github.com/Geoportail-Luxembourg/geoportailv3/commit/035de28ce4502b94c827fe3fa1f7bd709d6ad468

As a result, the coordinates display in the different projections only works properly in debug mode (where no variables are renamed). I have tried to convert to bracket notation but to no avail. 

Can someone @pgiraud / @elemoine point out what is going wrong here ?

pgiraud commented:
In your case, you forgot to use strings when you create the `mouseposition` property:
```javascript
this['mouseposition'] = new ol.control.MousePosition(...)
```

elemoine commented on geoportailv3/static/js/infobar/elevationdirective.js:
```diff
@@ -0,0 +1,70 @@
+    controllerAs: 'ctrl',
+    bindToController: true,
+    template: '<span class="elevation" translate>' +
+                'Elevation: {{elevation}} m</span>'
```
You use `controllerAs: 'ctrl'` (which is goog), so you want to use `{{ctrl.elevation}}` in the template. This requires placing `elevation` on the controller instead of directly on the scope.

elemoine commented on geoportailv3/static/js/infobar/elevationdirective.js:
```diff
@@ -0,0 +1,70 @@
+ * @constructor
+ */
+app.ElevationDirectiveController = function($scope, $http, ngeoDebounce) {
+  this['map'].on('pointermove',
```
We don't want to trigger requests when the map is panned. So `ngeoDebounce` should only be called when `e.dragging` is false:

```js
this['map'].on('pointermove',
    /**
     * @param {ol.MapBrowserEvent} e Map browser event.
     */
    function(e) {
      if (e.dragging) {
        return;
      }
      ngeoDebounce(…);
    });
```

elemoine commented on geoportailv3/static/js/infobar/elevationdirective.js:
```diff
@@ -0,0 +1,70 @@
+ */
+app.ElevationDirectiveController = function($scope, $http, ngeoDebounce) {
+  this['map'].on('pointermove',
+      ngeoDebounce(function(e) {
```
It would be good to type the callback function argument.

elemoine commented on geoportailv3/static/js/infobar/elevationdirective.js:
```diff
@@ -0,0 +1,70 @@
+              } else {
+                $scope['elevation'] = 'N/A';
+              }
+            });
```
So, to store the elevation on the controller, you can do as follows:

```js
    success(goog.bind(function(data) {
      this['elevation'] = data['dhm'] > 0 ? parseInt(data['dhm'] / 100, 0) : 'N/A';
    }, this));
```

elemoine commented on geoportailv3/static/js/infobar/infobardirective.js:
```diff
@@ -0,0 +1,49 @@
+ * @ngInject
+ * @constructor
+ */
+app.InfobarDirectiveController = function($scope, $document) {
```
You have unused arguments.

elemoine commented on geoportailv3/static/js/infobar/projectionselector.html:
```diff
@@ -0,0 +1,6 @@
+ <select ng-model="ctrl.projection" 
+                ng-options="opt as opt.label for opt in ctrl.projectionOptions" ng-change="ctrl.switchProjection()"></select>
```
You can use one-time binding for `ctrl.projectionOptions`.

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,124 @@
+                   return ol.coordinate.format(coord, '{x} | {y} (UTM31N)', 0);
+    }
+};
+/**
```
You need two empty lines above this line.

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,124 @@
+        projection = ol.proj.get(this['projection']['value']);
+    }
+    var widget = this.mouseposition;
+    console.log(projection);
```
`console.log` to be removed.

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,124 @@
+    console.log(projection);
+    widget.setProjection(projection);
+    widget.setCoordinateFormat(
+       /** @type {ol.CoordinateFormatType} */ (angular.bind(this, function(coord) {
```
We use `gooo.bind` instead of `angular.bind`, as `angular.bind` won't be renamed by the compiler. In general, we use `goog` function when available.

elemoine commented on geoportailv3/templates/index.html:
```diff
@@ -196,7 +197,9 @@ <h2 translate>infos</h2>
     <script src="${request.static_url('%s/ngeo/utils/watchwatchers.js' % node_modules_path)}"></script>
+    <script src="${request.static_url('%s/proj4/dist/proj4-src.js' % node_modules_path)}"></script>
 % else:
+    <script src="${request.static_url('%s/proj4/dist/proj4-src.js' % node_modules_path)}"></script>
```
Instead, we want to add proj4 to `build.js`. See how it's done in the Makefile.

elemoine replied:
As I said already, we want to add proj4 to `build.js`. But we can do that with a separate PR if you do not feel comfortable with it.

rmichaelis replied:
in issue #250 , you said that "proj4 should be used as an external library. It should not be compiled with the application" 
Why is it different now ?

elemoine replied:
It is not :-)

The `build.js` file located in `geoportailv3/static/build`, and loaded by the page in non-debug mode, is a concatenation of all the js builds. More specifically it's a concatenation of `jquery.min.js`, `angular.min.js`, `angular-animate.min.js`, `angular-gettext.min.js`, `bootstrap.min.js`, and `.build/build.js`, which is the app + ol3 + ngeo + closure Closure-compiled code. And soon, it will also include the application's directive partials (https://github.com/Geoportail-Luxembourg/geoportailv3/pull/293). I hope this clarifies.

rmichaelis replied:
Yes it clarifies :-) thanks

petzlux replied:
Can I take you up on the offer and merge this as is, and we deal with build.js in separate pull request.

elemoine replied:
Yes, that was my suggestion as well.

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,124 @@
+        this.mouseposition.setProjection(projection);
+        epsg_code = projection.getCode();
+    }
+    console.log(epsg_code);
```
`console.log` to be removed.

petzlux commented:
I managed to resolve the issues last Friday, forgot to remove the tags. 

@elemoine thanks, will take into account code comments

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,124 @@
+      return 'EPSG:32632';
+    } else {
+      return 'EPSG:32631';
+    }
```
How about:

```js
  return Math.floor(lonlat[0]) >= 6 ? 'EPSG:32632' : 'EPSG:32631';
```

?

elemoine commented on geoportailv3/static/js/infobar/projectionselector.html:
```diff
@@ -0,0 +1,6 @@
+ <select ng-model="ctrl.projection" 
+                ng-options="opt as opt.label for opt in ctrl.projectionOptions" ng-change="ctrl.switchProjection()"></select>
+<div class="mouse-coordinates"></div>
```
You could use the `ngeo-control` directive here.

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,124 @@
+      return this.coordFormat_(coord, this['projection']['value']);
+    }))
+  });
+  this.map.addControl(this.mouseposition) ;
```
As indicated above you could rely on the `ngeo-control` directive. In that case you just need to make the mouse position control available in the controller:

```js
  this['mousePositionControl'] = new ol.control.MousePosition({
    // …
  });
```

The `ngeo-control` directive will take care of adding the control to the map.

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,124 @@
+  this['projection'] = this['projectionOptions'][0];
+  this.mouseposition = new ol.control.MousePosition({
+    className: 'custom-mouse-coordinates',
+    target: $document.find('.mouse-coordinates')[0],
```
The `target` option should be removed if you use `ngeo-control`, as suggested.

petzlux replied:
@elemoine Can you please explain in more detail how to make use of ngeo-control in this context. The example in ngeo repo didnt really help me.

elemoine replied:
Use the following in the partial (as indicated above):

```html
<div class="mouse-coordinates" ngeo-control="ctrl.mousePositionControl" ngeo-control-map="ctrl.map"></div>
```

and that in the controller constructor:

```js
  this['mousePositionControl'] = new ol.control.MousePosition({
    className: 'custom-mouse-coordinates',
    projection: this.projection['value'],
    coordinateFormat: …
  });
```

Note that there's no `target` anymore, and no call to `map.addControl`. That's taken care of by the ngeo-control directive.

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,124 @@
+    className: 'custom-mouse-coordinates',
+    target: $document.find('.mouse-coordinates')[0],
+    projection: this.projection['value'],
+    coordinateFormat: /** @type {ol.CoordinateFormatType} */ (angular.bind(this, function(coord) {
```
Use `goog.bind`.

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,124 @@
+app.ProjectionselectorDirectiveController.prototype.coordFormat_ = function(coord, epsg_code)
+{
+    if (epsg_code === 'EPSG:3263*') {
+        var projection = ol.proj.get(this.utmZoneCheck_(coord));
```
Indent is incorrect. It is strange that this is not caught by the linters.

elemoine commented on geoportailv3/static/js/infobar/projectionselector.html:
```diff
@@ -0,0 +1,6 @@
+ <select ng-model="ctrl.projection" 
+                ng-options="opt as opt.label for opt in ::ctrl.projectionOptions" ng-change="ctrl.switchProjection()"></select>
+<div class="mouse-coordinates"></div>
```
To use `ngeo-control` you will need the following here for the "mouse-coordinates" div:

```html
<div class="mouse-coordinates" ngeo-control="ctrl.mousePositionControl" ngeo-control-map="ctrl.map"></div>
```

elemoine commented on geoportailv3/static/js/infobar/infobardirective.js:
```diff
@@ -0,0 +1,66 @@
+    this['infobarOpen'] = false;
+  } else {
+    this['infobarOpen'] = true;
+  }
```
Can be changed to

```js
this['infobarOpen'] = !this['infobarOpen'];
```

elemoine commented on geoportailv3/static/js/infobar/projectionselector.html:
```diff
@@ -0,0 +1,11 @@
+<div ng-class="::{'dropdown': true, 'dropup': true}">
```
Any reason for using `ng-class` here? May the classes change?

elemoine replied:
And any reason for having both `dropdown` and `dropup`?

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,150 @@
+  }
+  this['projection'] = goog.array.find(this['projectionOptions'],
+      function(obj) {
+        return obj.value == epsg_code;
```
You probably want to use `obj['value']` here.

elemoine commented:
I've added a few more comments. Otherwise it looks very good to me!

elemoine commented on geoportailv3/static/js/infobar/projectionselectordirective.js:
```diff
@@ -0,0 +1,150 @@
+ * @return {string}
+ */
+app.ProjectionselectorDirectiveController.prototype.coordFormat_ =
+    function(coord, epsg_code) {
```
Why not use `epsgCode`?

elemoine commented on CONST_requirements.txt:
```diff
@@ -3,5 +3,5 @@
 -r CONST_versions.txt
 -e git+https://github.com/camptocamp/pyramid_closure#egg=pyramid_closure
--e git+https://github.com/Pylons/pyramid@1e02bbfc0df09259bf207112acf019c8dba44a90#egg=pyramid
+-e git+https://github.com/sbrunner/pyramid@c2cgeoportal#egg=pyramid
```
@sbrunner, do we really want to use that?

petzlux commented:
Ready to merge ? @elemoine

elemoine commented on geoportailv3/static/js/infobar/elevationdirective.js:
```diff
@@ -0,0 +1,69 @@
+ * used to insert Elevation information into the HTML page.
+ * Example:
+ *
+ * <app-elevation app-elevation-map="::mainCtrl.map"map></app-elevation>
```
There's a typo here.

petzlux replied:
done

elemoine commented:
> Ready to merge ? @elemoine

I just noticed a typo in one of the examples. Please merge when this is fixed. And we need to create GitHub tickets for the remaining issues.

elemoine commented:
@petzlux feel free to delete the `info_bar` branch :-)
